pkg/apis/metrics/types.go

```diff
@@ -16,14 +16,427 @@ limitations under the License.
+	Network    NetworkStats `json:"network,omitempty"`
+	// Filesystem statistics
+	HasFilesystem bool      `json:"hasFilesystem,omitempty"`
+	Filesystem    []FsStats `json:"filesystem,omitempty"`
+	// Task load statistics
+	HasLoad bool      `json:"hasLoad,omitempty"`
+	Load    LoadStats `json:"load,omitempty"`
+	// Custom Metrics
+	HasCustomMetrics bool           `json:"hasCustomMetrics,omitempty"`
+	CustomMetrics    []CustomMetric `json:"customMetrics,omitempty"`
+}
+
+type TcpStat struct {
+	// Number of connections in the Established state.
+	Established uint64 `json:"established,omitempty"`
+	// Number of connections in the SynSent state.
+	SynSent uint64 `json:"synSent,omitempty"`
+	// Number of connections in the SynRecv state.
+	SynRecv uint64 `json:"synRecv,omitempty"`
+	// Number of connections in the FinWait1 state.
+	FinWait1 uint64 `json:"finWait1,omitempty"`
+	// Number of connections in the FinWait2 state.
+	FinWait2 uint64 `json:"finWait2,omitempty"`
+	// Number of connections in the TimeWait state.
+	TimeWait uint64 `json:"timeWait,omitempty"`
+	// Number of connections in the Close state.
+	Close uint64 `json:"close,omitempty"`
+	// Number of connections in the CloseWait state.
+	CloseWait uint64 `json:"closeWait,omitempty"`
+	// Number of connections in the LastAck state.
+	LastAck uint64 `json:"lastAck,omitempty"`
+	// Number of connections in the Listen state.
+	Listen uint64 `json:"listen,omitempty"`
+	// Number of connections in the Closing state.
+	Closing uint64 `json:"closing,omitempty"`
+}
+
+type NetworkStats struct {
+	// Network stats by interface.
+	Interfaces []InterfaceStats `json:"interfaces,omitempty"`
+	// TCP connection stats (Established, Listen...)
+	Tcp TcpStat `json:"tcp,omitempty"`
+	// TCP6 connection stats (Established, Listen...)
+	Tcp6 TcpStat `json:"tcp6,omitempty"`
+}
+
+type InterfaceStats struct {
+	// The name of the interface.
+	Name string `json:"name,omitempty"`
+	// Cumulative count of bytes received.
+	RxBytes uint64 `json:"rxBytes,omitempty"`
+	// Cumulative count of packets received.
+	RxPackets uint64 `json:"rxPackets,omitempty"`
+	// Cumulative count of receive errors encountered.
+	RxErrors uint64 `json:"rxErrors,omitempty"`
+	// Cumulative count of packets dropped while receiving.
+	RxDropped uint64 `json:"rxDropped,omitempty"`
+	// Cumulative count of bytes transmitted.
+	TxBytes uint64 `json:"txBytes,omitempty"`
+	// Cumulative count of packets transmitted.
+	TxPackets uint64 `json:"txPackets,omitempty"`
+	// Cumulative count of transmit errors encountered.
+	TxErrors uint64 `json:"txErrors,omitempty"`
+	// Cumulative count of packets dropped while transmitting.
+	TxDropped uint64 `json:"txDropped,omitempty"`
+}
+
+// Instantaneous CPU stats
+type CpuInstStats struct {
+	Usage CpuInstUsage `json:"usage,omitempty"`
+}
+
+// CPU usage time statistics.
+type CpuInstUsage struct {
+	// Total CPU usage.
+	// Units: nanocores per second
```

Comment:
Yeah. That makes sense. Maybe `cpu-nanoseconds / second`?

On Thu, Nov 12, 2015 at 1:36 PM, Tim St. Clair <notifications@github.com>
wrote:

> In pkg/apis/metrics/types.go
> <https://github.com/kubernetes/kubernetes/pull/16458#discussion_r44718079>
> :
>
> > +	TxPackets uint64 `json:"txPackets,omitempty"`
> > +	// Cumulative count of transmit errors encountered.
> > +	TxErrors uint64 `json:"txErrors,omitempty"`
> > +	// Cumulative count of packets dropped while transmitting.
> > +	TxDropped uint64 `json:"txDropped,omitempty"`
> > +}
> > +
> > +// Instantaneous CPU stats
> > +type CpuInstStats struct {
> > +	Usage CpuInstUsage `json:"usage,omitempty"`
> > +}
> > +
> > +// CPU usage time statistics.
> > +type CpuInstUsage struct {
> > +	// Total CPU usage.
> > +	// Units: nanocores per second
>
> I think Daniel's points are valid. It doesn't make sense to use 1 cpu *in*
> one second, you would use 1 cpu *for* 1 second (1 cpu-second), so then
> the "instantaneous" (which is really a 10-second rolling average) usage
> would be N cpu-seconds / second, which I think is N*10^9 nanocore-seconds /
> second? (or does a nanocore include seconds?)
>
> I think we've concluded that changing the units is out-of-scope for this
> PR, but I do want to at least get the units documented correctly.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/16458/files#r44718079>.
>